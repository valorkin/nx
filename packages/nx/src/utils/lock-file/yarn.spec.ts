import {
  parseYarnLockFile,
  pruneYarnLockFile,
  stringifyYarnLockFile,
} from './yarn';
import {
  lockFile,
  berryLockFile,
  lockFileJustTypescript,
  lockFileDevkitAndYargs,
  berryLockFileDevkitAndYargs,
} from './__fixtures__/yarn.lock';

describe('yarn LockFile utility', () => {
  describe('classic', () => {
    const parsedLockFile = parseYarnLockFile(lockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toBeUndefined();
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(324);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
    });

    it('should map various instances of the same version', () => {
      const babelCoreDependency =
        parsedLockFile.dependencies['@babel/core']['@babel/core@7.19.1'];
      expect(babelCoreDependency.packageMeta.length).toEqual(2);
      expect(babelCoreDependency.packageMeta).toEqual([
        '@babel/core@^7.11.6',
        '@babel/core@^7.12.3',
      ]);
    });

    it('should match the original file on stringification', () => {
      expect(stringifyYarnLockFile(parsedLockFile)).toEqual(lockFile);
    });

    it('shold prune the lock file', () => {
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, ['typescript']).dependencies
        ).length
      ).toEqual(1);
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, ['yargs', '@nrwl/devkit'])
            .dependencies
        ).length
      ).toEqual(36);
    });

    it('shold correctly prune lockfile with single package', () => {
      expect(
        stringifyYarnLockFile(pruneYarnLockFile(parsedLockFile, ['typescript']))
      ).toEqual(lockFileJustTypescript);
    });

    it('shold correctly prune lockfile with multiple packages', () => {
      expect(
        stringifyYarnLockFile(
          pruneYarnLockFile(parsedLockFile, ['yargs', '@nrwl/devkit'])
        )
      ).toEqual(lockFileDevkitAndYargs);
    });
  });

  describe('berry', () => {
    const parsedLockFile = parseYarnLockFile(berryLockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({
        __metadata: { cacheKey: '8', version: '6' },
        workspacePackages: {
          'test@workspace:.': {
            dependencies: {
              '@nrwl/cli': '14.7.5',
              '@nrwl/workspace': '14.7.5',
              nx: '14.7.5',
              prettier: '^2.6.2',
              typescript: '~4.8.2',
            },
            languageName: 'unknown',
            linkType: 'soft',
            resolution: 'test@workspace:.',
            version: '0.0.0-use.local',
          },
        },
      });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(386);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
    });

    it('should map various instances of the same version', () => {
      const babelCoreDependency =
        parsedLockFile.dependencies['@babel/core']['@babel/core@7.19.1'];
      expect(babelCoreDependency.packageMeta.length).toEqual(2);
      expect(babelCoreDependency.packageMeta).toEqual([
        '@babel/core@npm:^7.11.6',
        '@babel/core@npm:^7.12.3',
      ]);
    });

    it('should match the original file on stringification', () => {
      const result = stringifyYarnLockFile(parsedLockFile);
      expect(result).toMatch(
        /This file was generated by Nx. Do not edit this file directly/
      );
      expect(removeComment(result)).toEqual(removeComment(berryLockFile));
    });

    it('shold prune the lock file', () => {
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, [
            'yargs',
            '@nrwl/devkit',
            'typescript',
          ]).dependencies
        ).length
      ).toEqual(37);
    });

    it('shold correctly prune lockfile with multiple packages', () => {
      const result = stringifyYarnLockFile(
        pruneYarnLockFile(parsedLockFile, [
          'yargs',
          '@nrwl/devkit',
          'typescript',
        ])
      );
      expect(removeComment(result)).toEqual(
        removeComment(berryLockFileDevkitAndYargs)
      );
    });
  });
});

// we don't care about comment message
const removeComment = (value) => value.split(/\n/).slice(2).join('\n');
