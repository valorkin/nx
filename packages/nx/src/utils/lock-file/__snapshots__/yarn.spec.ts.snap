// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yarn LockFile utility berry should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "checksum": "d74d170d06468913921d72430259424b7e4c826b5a7d39ff839a29d547efb97dc577caa8ba3fb5cf023624e9af9d09651afc3d4112a45e2050328abc9b3a2292",
    "dependencies": Object {
      "@jridgewell/gen-mapping": "^0.1.0",
      "@jridgewell/trace-mapping": "^0.3.9",
    },
    "languageName": "node",
    "linkType": "hard",
    "packageMeta": Array [
      "@ampproject/remapping@npm:^2.1.0",
    ],
    "resolution": "@ampproject/remapping@npm:2.2.0",
    "version": "2.2.0",
  },
}
`;

exports[`yarn LockFile utility berry should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.3": Object {
    "bin": Object {
      "tsc": "bin/tsc",
      "tsserver": "bin/tsserver",
    },
    "checksum": "8286a5edcaf3d68e65c451aa1e7150ad1cf53ee0813c07ec35b7abdfdb10f355ecaa13c6a226a694ae7a67785fd7eeebf89f845da0b4f7e4a35561ddc459aba0",
    "languageName": "node",
    "linkType": "hard",
    "packageMeta": Array [
      "typescript@npm:~4.8.2",
    ],
    "resolution": "typescript@npm:4.8.3",
    "version": "4.8.3",
  },
}
`;

exports[`yarn LockFile utility classic should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "dependencies": Object {
      "@jridgewell/gen-mapping": "^0.1.0",
      "@jridgewell/trace-mapping": "^0.3.9",
    },
    "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    "packageMeta": Array [
      "@ampproject/remapping@^2.1.0",
    ],
    "resolved": "https://registry.yarnpkg.com/@ampproject/remapping/-/remapping-2.2.0.tgz#56c133824780de3174aed5ab6834f3026790154d",
    "version": "2.2.0",
  },
}
`;

exports[`yarn LockFile utility classic should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.3": Object {
    "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    "packageMeta": Array [
      "typescript@~4.8.2",
    ],
    "resolved": "https://registry.yarnpkg.com/typescript/-/typescript-4.8.3.tgz#d59344522c4bc464a65a730ac695007fdb66dd88",
    "version": "4.8.3",
  },
}
`;
